src/connection-loop.ts

```diff
@@ -0,0 +1,261 @@
+
+      // This resolver is used to wait for incoming push calls.
+      this._pendingResolver = resolver();
+
+      if (counter >= maxConnections()) {
+        log?.('Too many pushes. Waiting until one finishes...');
+        await this._waitUntilAvailableConnection();
+        log?.('...finished');
+      }
+
+      // We need to delay the next request even if there are no active requests
+      // in case of error.
+      if (counter > 0 || didLastSendRequestFail(sendRecords)) {
+        delay = computeDelayAndUpdateDurations(
+          delay,
+          maxConnections(),
+          sendRecords,
+        );
+        log?.(
+          didLastSendRequestFail(sendRecords)
+            ? 'Last connection errored. Sleeping for'
+            : 'More than one outstanding connection (' +
+                counter +
+                '). Sleeping for',
+          delay,
+          'ms',
+        );
+
+        const timeSinceLastSend = Date.now() - lastSendTime;
+        if (delay > timeSinceLastSend) {
+          await Promise.race([
+            sleep(delay - timeSinceLastSend),
+            recoverResolver.promise,
+          ]);
+        }
+      }
+
+      counter++;
+      (async () => {
+        const start = Date.now();
+        let err: unknown;
+        try {
+          lastSendTime = start;
+          await delegate.invokeSend();
+        } catch (e) {
+          err = e;
+        }
+        sendRecords.push({duration: Date.now() - start, ok: !err});
+        if (recovered(sendRecords)) {
+          recoverResolver.resolve();
+          recoverResolver = resolver();
+        }
+        counter--;
+        this._connectionAvailable();
+        if (err) {
+          currentResolver.reject(err);
+          // Keep trying
+          this._pendingResolver.resolve();
+        } else {
+          currentResolver.resolve();
+        }
+      })();
+    }
+  }
+
+  private _waitingConnectionResolve: (() => void) | undefined = undefined;
+
+  private _connectionAvailable() {
+    if (this._waitingConnectionResolve) {
+      const resolve = this._waitingConnectionResolve;
+      this._waitingConnectionResolve = undefined;
+      resolve();
+    }
+  }
+
+  private _waitUntilAvailableConnection() {
+    const {promise, resolve} = resolver();
+    this._waitingConnectionResolve = resolve;
+    return promise;
+  }
+}
+
+// Number of connections to remember when computing the new delay.
+const CONNECTION_MEMORY_COUNT = 9;
+
+function computeDelayAndUpdateDurations(
+  delay: number,
+  maxConnections: number,
+  sendRecords: SendRecord[],
+): number {
+  const {length} = sendRecords;
+  if (length === 0) {
+    return delay;
+  }
+
+  const {duration, ok} = sendRecords[sendRecords.length - 1];
+
+  if (!ok) {
+    return Math.min(MAX_DELAY, delay * 2);
+  }
+
+  if (length === 1) {
+    return (duration / maxConnections) | 0;
+  }
+
+  // length > 1
+  const previous: SendRecord = sendRecords[sendRecords.length - 2];
+
+  // Prune
```

Comment:
Here it removes a single record, but is it possible for sendrecords to grow in length such that we should be removing more than one? I ask because line 214 returns early in the error case and it seems like in that case we keep adding sendrecords while it is erroring, resulting in many more records than we wish to keep. Perhaps we should remove extra records in the main connection loop or at the top of this function, and remove as many as required to reduce the length to CONNECTION_MEMORY_COUNT?

Reply:
Good observation. It does seem possible.